clash-cores/src/Clash/Cores/Crc/Internal.hs

```diff
@@ -0,0 +1,737 @@
+reset
+  :: forall (crcWidth :: Nat) (dataWidth :: Nat)
+   . SoftwareCrc crcWidth dataWidth
+  -- ^ Current CRC engine
+  -> SoftwareCrc crcWidth dataWidth
+  -- ^ CRC engine fresh to be used
+reset engine@(SoftwareCrc {..}) = go _crcDataWidth _crcParams
+  where
+    CrcParams {..} = _crcParams
+    go SNat (CrcParams SNat _ _ _ _ _) = engine { _crcCurrent = initialCrc }
+      where
+        initial :: BitVector (crcWidth + dataWidth)
+        initial = extend _crcInitial
+        initialCrc = shiftL initial (snatToNum _crcDataWidth)
+
+-- | Feed a word to the 'SoftwareCrc'
+feed
+  :: forall (crcWidth :: Nat) (dataWidth :: Nat)
+   . SoftwareCrc crcWidth dataWidth
+  -- ^ Current 'SoftwareCrc'
+  -> BitVector dataWidth
+  -- ^ Update 'SoftwareCrc' with this input
+  -> SoftwareCrc crcWidth dataWidth
+  -- ^ 'SoftwareCrc' with update processed
+feed engine@(SoftwareCrc {..}) word = go _crcDataWidth _crcParams
+  where
+    CrcParams {..} = _crcParams
+    go SNat (CrcParams SNat _ _ _ _ _) = engine { _crcCurrent = nextCrc }
+      where
+        applyPolyBit crcB
+          | crcB .&. _crcTopBitMask > 0 = (shiftL crcB 1) `xor` _crcpolyShifted
+          | otherwise                   = shiftL crcB 1
+        applyPoly crcW = L.iterate applyPolyBit crcW L.!! snatToNum _crcDataWidth
+        applyCrc crc w = applyPoly $ crc `xor` shiftL (extend w) (snatToNum _crcWidth)
+        nextCrc = applyCrc _crcCurrent $ applyWhen _crcReflectInput reverseBV word
+
+-- | Get current CRC state. This is the value before any output reflection and output XOR.
+getCrcState
+  :: forall (crcWidth :: Nat) (dataWidth :: Nat)
+   . SoftwareCrc crcWidth dataWidth
+  -> BitVector crcWidth
+getCrcState (SoftwareCrc {..}) = go _crcDataWidth _crcParams
+  where
+    go SNat (CrcParams SNat _ _ _ _ _) = fst $ split _crcCurrent
+
+-- | Generate the CRC
+--
+--   If the 'SoftwareCrc' will be reused, 'reset' must be called before starting
+--   to process the next message.
+digest
+  :: forall (crcWidth :: Nat) (dataWidth :: Nat)
+   . SoftwareCrc crcWidth dataWidth
+  -> BitVector crcWidth
+digest params@(SoftwareCrc {..}) = go _crcDataWidth _crcParams
+  where
+    go SNat (CrcParams SNat _ _ _ _ _) = outputCrcRev `xor` _crcXorOutput
+      where
+        CrcParams {..} = _crcParams
+        outputCrcRev = applyWhen _crcReflectOutput reverseBV $ getCrcState params
+
+-- | Compute the raw residue value for these 'CrcParams', which is the value left in the
+--   CRC register after processing any valid codeword. Raw in this case means
+--   this is the residue before the output reflection and out xor has been applied.
+--
+--   A fixed amount of padding can be given. This is useful during CRC verification
+--   in the scenario where where you want to handle more words in a single cycle
+--   than are native to the protocol.
+--
+--   For example ethernet is byte-oriented but you may want to handle 32-bit
+--   in a single cycle to achieve a high throughput.
```

Comment:
```suggestion
--   For example ethernet is byte-oriented but you may want to handle 32 bit
--   in a single cycle to achieve high throughput.
```

Dropping the article was already part of a previous review comment with suggestion; perhaps the "commit suggestion" function of GitHub is a nice way to apply suggestions so you don't run the risk of accidentally omitting part of the suggestion?